modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/snapshot/DoubleSignDetect.scala

```diff
@@ -0,0 +1,81 @@
+package org.tessellation.node.shared.domain.snapshot
+
+import cats.effect.Async
+import cats.effect.std.Supervisor
+import cats.syntax.all._
+
+import org.tessellation.node.shared.config.types.{DoubleSignDetectConfig, DoubleSignDetectDaemonConfig}
+import org.tessellation.node.shared.domain.Daemon
+import org.tessellation.node.shared.domain.fork.{ForkInfo, ForkInfoEntries, ForkInfoStorage}
+import org.tessellation.node.shared.domain.snapshot.DoubleSignDetect.IndexedForkInfo
+import org.tessellation.schema.peer.PeerId
+
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.auto._
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.types.numeric.NonNegLong
+import org.typelevel.log4cats.SelfAwareStructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object DoubleSignDetect {
+  case class IndexedForkInfo(info: ForkInfo, idx: NonNegLong)
+
+  def daemon[F[_]: Async: Supervisor](
+    doubleSignDetect: DoubleSignDetect[F],
+    config: DoubleSignDetectDaemonConfig
+  ): Daemon[F] =
+    Daemon.periodic(doubleSignDetect.run, config.runInterval)
+}
+
+class DoubleSignDetect[F[_]: Async](
+  storage: ForkInfoStorage[F],
+  config: DoubleSignDetectConfig
+) {
+  private val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger
+
+  private def isDoubleSign(previous: IndexedForkInfo, current: IndexedForkInfo): Boolean =
+    (current.idx - previous.idx <= config.maxDetectionDelta) && (previous.info.hash =!= current.info.hash)
```

Comment:
Also consider such a case that node is forked and sends you proposal `AAA`  for ordinal `100` and then rejoins and sends another proposal `BBB` for ordinal `100`. You end up with a false-positive and you treat such node as malicious. For such a case the observing state probably helps as node "waits" with participating in consensus after the rejoin.

Reply:
> I am also wondering why `<= config.maxDetectionDelta` is considered as a double sign but `> maxDetectionDelta` is not 🤔 . What is the reasoning behind considering one delta as "safe" and another as "unsafe"?

Regarding this, @kpudlik , the reasoning is that if a node does a rollback, it will also transmit different snapshot hashes for the same ordinals. For example, it sends hash `BB` for ordinal `100` when it previously sent hash `AA` for that ordinal. The reasoning is that rollbacks will cause conflicting hashes to be detected, but they will be beyond `maxDetectionDelta`. So, we care about conflicts when they are within `maxDetectionDelta`.

Reply:
>So the functionality is as following:
>Node signs snapshot of hash AAA at ordinal 100 and then within >maxDetectionDelta signs snapshot of hash BBB also at ordinal >100 and such behavior is considered as a double-sign. I doubt it is >enough to consider it a double sign because lastSnapshotHash >may differ.

It is enough, because we’re assuming here that somewhere else in the ordinal that lastSnapshotHash will be treated as another hash in the iterable, same logic we’re using elsewhere to determine a single hash for ForkInfo as being different constituting a fork.

> The only case when the double sign is obvious is when:
> node proposed 2 proposals or more
> lastSnapshotHashes are equal
> ordinals are equal
> all the proposals were sent during the same consensus round (but > here check happens every config.runInterval for hardcoded delta > which may consider the proposal sent outside the consensus > round as double-sign)

This has already been addressed earlier, the main problem with considering only the proposals of the current round as opposed to a time series window is we’re currently dealing with a large number of honest forks due to unrelated issue which has not yet been fixed. We can’t aggressively penalize a behavior due to a separate underlying problem, hence the use of a more strict check here over a longer period of time.

> I am also wondering why <= config.maxDetectionDelta is > >
> considered as a double sign but > maxDetectionDelta is not 🤔 . 

Reference the earlier function before where this is invoked, the delta is on the gossip receipt index, NOT the ordinal. We’re grouping by the ordinal. The distinction here we’re trying to detect is whether or not we’re receiving similar ordinals at similar receipt order, which is NOT the case that happens during a rollback.

> What is the reasoning behind considering one delta as "safe" and > another as "unsafe"? It can cause "faster" forks to be treated as > malicious but "slower" ones as correct. Relying on magic number > and some "distance" is risky.

Again, this is the wrong interpretation, because we’re doing a groupBy ordinal and delta on receipt order. NOT a delta on ordinal, hence it’s not detecting a ‘faster’ fork’, but rather a concurrently repeating double fork.

I still need to review more to determine if the rest of this is correctly implemented, but in plain English we’re attempting to find similar ordinals with differing hashes that are received in close gossip order. Ideally we’d be expecting some calculation like this to occur over multiple iterations, but for now I think it’s just set to detect a single event. A better (future) implementation would involve detecting this over a longer period of time.


>Also even if we find the perfect maxDetectionDelta and we check > the lastSnapshotHash as well then it still doesn't solve the > double-sign problem. 

This PR doesn’t attempt to solve every case, just one to start with, see earlier comments regarding how we can later make improvements to it.

> Consider such an example:
> A malicious node creates a proposal of hash BBB at ordinal 100 and creates a proposal of hash AAA at ordinal 100. Then he can either:
>”send both, sign one"

This has been addressed earlier, we can derive the information from the signed artifacts which can be done as a separate PR.

> “sign both, send both"

Detected by forkInfo


> and additionally
> “lie in fork-info about which one has been signed"

We’re using forkInfo to detect forks and reset the node against it (or will with nodectl), so they’ll ignore the signed artifact if we don’t include them in that calculation.

> “send many fork-infos to reach the max-size of storage"

Part of a DDOS related problem that can be addressed as part of efforts to eliminate DDOS / spam, this is not the only place where this is an issue, but actually Chuong this is an appropriate place where we could impose a per-node history limit (need to check if that’s already in here.)


>Let's say that such a node decided that BBB is proposed during a proposal round and then lied that he proposed AAA in a fork-info. >Or even didn't not send the fork-info which is completely legal as you do not check if everyone sent it.

We’re using it as a basis for a check which can potentially reset the node, so it’s used.

> So basically the problem with your solution is that each node sends 2 things independently: Proposal and ForkInfo where in theory both should match but in reality malicious node can send the malformed ForkInfo or not send it at all and nothing will happen because gossiping is a step after the consensus.

I’ve already responded to this in detail in Maciej’s comment, these issues can be treated independently and we can derive the exact same info from the other step.


> I think it is the same as Maciek said:
The fork info declarations are sent by the node in question itself. Hence the malicious node can decide to simply not send any declarations and having that we will not detect anything. 

Again, just repeating my earlier comments here, if we’re using it to reset the node to the correct fork, it’s fundamentally a requirement, even then on a shorter term basis, we’ll change the derivation source of the information so it’s derived from the signatures, which is a separate change.

> Even if the malicious node decided to send declarations it could send them in such a way that would let it manipulate the result on the nodes checking them. That results from the reliance on the order of declarations put into the storage and the storage capacity itself.

Any node switching chains still needs to validate the underlying data, this is not any different from saying you can ‘manipulate’ the active fork by a majority operation so long as it’s valid, that’s always a concern. The storage problem I agree with should potentially be solved in here to limit the amount of info per node.

> That's why I think that instead of having a separate time-based step that relies on magic distance we should just run the detection as a part of the consensus round 

We can’t do that as already explained, considering how many honest short-time forks we see it’d penalize nodes that are absolutely not malicious until that is solved separately.

> (so you get sync with consensus for free). Currently, consensus only stores the very first proposal sent by the node and all next proposals (which may be double-sign) are just ignored. 

I agree that should be fixed, but there’s no reason not to consider a time-window of this data to be of importance. We’ve discussed this earlier, and even if we stored a time window inside the consensus proposal classes, it would essentially be the exact same data model proposed in this, since we don’t care about the underlying data in the proposals, but rather just the hashes. So either way, it would recreate what we’re doing here, it’s just a matter of organization.

A later step (after solving some of the meaningless forking issues,) would in fact be to ALSO include the information about the latest proposal, to distinguish fork events that are immediate versus those over a time window.

> What we could do is just to store all the proposals sent by the node instead of the first one and then after collecting them just check if any of the nodes sent the proposal for the same ordinal and lastSnapshotHash twice or more. So instead of proposal: Option[Proposal], just store proposal: Option[NonEmptyList[Proposal]] and all the nodes who have more than 2 proposals for the same ordinal and lastSnapshotHash are double-signers.

Already addressed earlier, same problem as before where we get a ton of false positives.

> No relying on time -> no race conditions

We’re not relying on explicit time here, but receipt order.
 > No relying on distance -> no favoring slower forks and no need for hyperparameter optimization

In general, any detection function would be best considering a window of events, rather than the most immediate event. It’s useful to distinguish the severity of the malicious behavior, where a node acting over a longer period of events is considered ‘worse’ than an immediate operation, which could simply be due to connectivity issues or networking related problems.
 > Relying on consensus round -> sync with consensus and be enclosed within the context of a single round with access to all the data to detect double-sign

Again, the context of a single round is strictly speaking less information than a window of events over multiple rounds, which allows us to detect and distinguish between different types of situations.


> Also consider such a case that node is forked and sends you proposal AAA for ordinal 100 and then rejoins and sends another proposal BBB for ordinal 100 (the correct one this time because it is not forked anymore). You end up with a false-positive and you treat such node as malicious. For such a case the observing state probably helps as node "waits" with participating in consensus after the rejoin.

 Since we’re planning on resets through restarts, I don’t think this can happen easily. By the time the node resets, it would have NOT done a new proposal round for round 100, but would instead ‘rejoin’ at round 105 or so. It would redownload against the ‘correct’ chain, and then attempt proposals at the ‘later’ round. So actually I don’t think it would even happen, because of the way we reset nodes.

Reply:
Ok, I went through the answers and I think that that PR still doesn't completely cover the double-sign problem but may solve the forking problem. On the other hand, I agree that stricter mechanisms may lead to false penalties due to the forking issues we have. 
>This has already been addressed earlier, the main problem with considering only the proposals of the current round as opposed to a time series window is we’re currently dealing with a large number of honest forks due to unrelated issue which has not yet been fixed

^ this is also the reason why it needs to be implemented outside of single consensus round context.

Syntactically everything is ok, but I am still wondering if it works properly semantically so if you tested it and you are planning to create more PRs to cover more cases then feel free to merge that.

Reply:
I've created a separate ticket to create an attack harness to rigorously test double signing detection. For the purposes of this PR, the feature is only being used for logging. So, we only want to ensure that it doesn't break anything.